src/probability_theory/density.lean

```diff
@@ -0,0 +1,329 @@
+
+namespace measure_theory
+
+open topological_space
+
+variables {α E : Type*} [normed_group E] [measurable_space E] [second_countable_topology E]
+  [normed_space ℝ E] [complete_space E] [borel_space E]
+
+namespace measure
+
+/-- A random variable `X : α → E` is said to `has_pdf` with respect to the measure `ℙ` on `α` and
+`μ` on `E` if there exists a measurable function `f` such that the push-forward measure of `ℙ`
+along `X` equals `μ.with_density f`. -/
+class has_pdf {m : measurable_space α} (X : α → E)
+  (ℙ : measure α) (μ : measure E . volume_tac) : Prop :=
+(pdf' : measurable X ∧ ∃ (f : E → ℝ≥0∞), measurable f ∧ map X ℙ = μ.with_density f)
+
+@[measurability]
+lemma has_pdf.measurable {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ] :
+  measurable X :=
+hX.pdf'.1
+
+/-- If `X` is a random variable that `has_pdf X ℙ μ`, then `pdf X` is the measurable function `f`
+such that the push-forward measure of `ℙ` along `X` equals `μ.with_density f`. -/
+def pdf {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :=
+if hX : has_pdf X ℙ μ then classical.some hX.pdf'.2 else 0
+
+lemma pdf_undef {m : measurable_space α} {ℙ : measure α} {μ : measure E} {X : α → E}
+  (h : ¬ has_pdf X ℙ μ) :
+  pdf X ℙ μ = 0 :=
+by simp only [pdf, dif_neg h]
+
+@[measurability]
+lemma measurable_pdf {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :
+  measurable (pdf X ℙ μ) :=
+begin
+  by_cases hX : has_pdf X ℙ μ,
+  { rw [pdf, dif_pos hX],
+    exact (classical.some_spec hX.pdf'.2).1 },
+  { rw [pdf, dif_neg hX],
+    exact measurable_zero }
+end
+
+lemma map_eq_with_density_pdf {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ] :
+  measure.map X ℙ = μ.with_density (pdf X ℙ μ) :=
+begin
+  rw [pdf, dif_pos hX],
+  exact (classical.some_spec hX.pdf'.2).2
+end
+
+lemma map_eq_set_lintegral_pdf {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ]
+  {s : set E} (hs : measurable_set s) :
+  measure.map X ℙ s = ∫⁻ x in s, pdf X ℙ μ x ∂μ :=
+by rw [← with_density_apply _ hs, map_eq_with_density_pdf X ℙ μ]
+
+lemma pdf_ae_eq_zero_of_not_measurable {m : measurable_space α}
+  {ℙ : measure α} {μ : measure E} {X : α → E} (hX : ¬ measurable X) :
+  pdf X ℙ μ =ᵐ[μ] 0 :=
+begin
+  by_cases hX_pdf : has_pdf X ℙ μ,
+  { haveI := hX_pdf,
+    have h_eq := map_eq_with_density_pdf X ℙ μ,
+    rw map_of_not_measurable hX at h_eq,
+    exact with_density_eq_zero (measurable_pdf X ℙ μ).ae_measurable h_eq.symm, },
+  { exact filter.eventually_of_forall (λ x, by rw pdf_undef hX_pdf), },
+end
+
+lemma measurable_of_pdf_ne_zero {m : measurable_space α}
+  {ℙ : measure α} {μ : measure E} (X : α → E) (h : ¬ pdf X ℙ μ =ᵐ[μ] 0) :
+  measurable X :=
+by { by_contra hX, exact h (pdf_ae_eq_zero_of_not_measurable hX), }
```

Comment:
These two can now be changed to reflect that the pdf of a non-measurable function is 0 (exactly, not a.e.). And the case `has_pdf` can be dropped from the first proof.